File: Masjidly - Official Masjid Prayer Times/Features/Notifications/NotificationSettings.swift
import Foundation

struct NotificationSettings: Codable, Equatable, Sendable {
    var masterEnabled: Bool = false
    var adhanEnabled: Bool = true
    var iqamahEnabled: Bool = true
    var preAdhanReminderMinutes: Int? = nil
    var preIqamahReminderMinutes: Int? = nil
    // Legacy single per-prayer flags (used for both adhan + iqamah before v1.1.2)
    var fajr: Bool = true
    var dhuhrJummah: Bool = true
    var asr: Bool = true
    var maghrib: Bool = true
    var isha: Bool = true
    // Per-type per-prayer flags (v1.1.2+)
    var adhanFajr: Bool = true
    var adhanDhuhrJummah: Bool = true
    var adhanAsr: Bool = true
    var adhanMaghrib: Bool = true
    var adhanIsha: Bool = true
    var iqamahFajr: Bool = true
    var iqamahDhuhrJummah: Bool = true
    var iqamahAsr: Bool = true
    var iqamahMaghrib: Bool = true
    var iqamahIsha: Bool = true

    enum CodingKeys: String, CodingKey {
        case masterEnabled
        case adhanEnabled
        case iqamahEnabled
        case preAdhanReminderMinutes
        case preIqamahReminderMinutes
        case fajr
        case dhuhrJummah
        case asr
        case maghrib
        case isha
        case adhanFajr
        case adhanDhuhrJummah
        case adhanAsr
        case adhanMaghrib
        case adhanIsha
        case iqamahFajr
        case iqamahDhuhrJummah
        case iqamahAsr
        case iqamahMaghrib
        case iqamahIsha
    }

    init(
        masterEnabled: Bool = false,
        adhanEnabled: Bool = true,
        iqamahEnabled: Bool = true,
        preAdhanReminderMinutes: Int? = nil,
        preIqamahReminderMinutes: Int? = nil,
        fajr: Bool = true,
        dhuhrJummah: Bool = true,
        asr: Bool = true,
        maghrib: Bool = true,
        isha: Bool = true,
        adhanFajr: Bool = true,
        adhanDhuhrJummah: Bool = true,
        adhanAsr: Bool = true,
        adhanMaghrib: Bool = true,
        adhanIsha: Bool = true,
        iqamahFajr: Bool = true,
        iqamahDhuhrJummah: Bool = true,
        iqamahAsr: Bool = true,
        iqamahMaghrib: Bool = true,
        iqamahIsha: Bool = true
    ) {
        self.masterEnabled = masterEnabled
        self.adhanEnabled = adhanEnabled
        self.iqamahEnabled = iqamahEnabled
        self.preAdhanReminderMinutes = preAdhanReminderMinutes
        self.preIqamahReminderMinutes = preIqamahReminderMinutes
        self.fajr = fajr
        self.dhuhrJummah = dhuhrJummah
        self.asr = asr
        self.maghrib = maghrib
        self.isha = isha
        self.adhanFajr = adhanFajr
        self.adhanDhuhrJummah = adhanDhuhrJummah
        self.adhanAsr = adhanAsr
        self.adhanMaghrib = adhanMaghrib
        self.adhanIsha = adhanIsha
        self.iqamahFajr = iqamahFajr
        self.iqamahDhuhrJummah = iqamahDhuhrJummah
        self.iqamahAsr = iqamahAsr
        self.iqamahMaghrib = iqamahMaghrib
        self.iqamahIsha = iqamahIsha
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        masterEnabled = try c.decodeIfPresent(Bool.self, forKey: .masterEnabled) ?? false
        adhanEnabled = try c.decodeIfPresent(Bool.self, forKey: .adhanEnabled) ?? true
        iqamahEnabled = try c.decodeIfPresent(Bool.self, forKey: .iqamahEnabled) ?? true
        preAdhanReminderMinutes = try c.decodeIfPresent(Int.self, forKey: .preAdhanReminderMinutes)
        preIqamahReminderMinutes = try c.decodeIfPresent(Int.self, forKey: .preIqamahReminderMinutes)
        // Legacy single per-prayer flags
        fajr = try c.decodeIfPresent(Bool.self, forKey: .fajr) ?? true
        dhuhrJummah = try c.decodeIfPresent(Bool.self, forKey: .dhuhrJummah) ?? true
        asr = try c.decodeIfPresent(Bool.self, forKey: .asr) ?? true
        maghrib = try c.decodeIfPresent(Bool.self, forKey: .maghrib) ?? true
        isha = try c.decodeIfPresent(Bool.self, forKey: .isha) ?? true
        // Per-type per-prayer flags: fall back to legacy flag if missing
        adhanFajr = try c.decodeIfPresent(Bool.self, forKey: .adhanFajr) ?? fajr
        adhanDhuhrJummah = try c.decodeIfPresent(Bool.self, forKey: .adhanDhuhrJummah) ?? dhuhrJummah
        adhanAsr = try c.decodeIfPresent(Bool.self, forKey: .adhanAsr) ?? asr
        adhanMaghrib = try c.decodeIfPresent(Bool.self, forKey: .adhanMaghrib) ?? maghrib
        adhanIsha = try c.decodeIfPresent(Bool.self, forKey: .adhanIsha) ?? isha
        iqamahFajr = try c.decodeIfPresent(Bool.self, forKey: .iqamahFajr) ?? fajr
        iqamahDhuhrJummah = try c.decodeIfPresent(Bool.self, forKey: .iqamahDhuhrJummah) ?? dhuhrJummah
        iqamahAsr = try c.decodeIfPresent(Bool.self, forKey: .iqamahAsr) ?? asr
        iqamahMaghrib = try c.decodeIfPresent(Bool.self, forKey: .iqamahMaghrib) ?? maghrib
        iqamahIsha = try c.decodeIfPresent(Bool.self, forKey: .iqamahIsha) ?? isha
    }
}

File: apps/android/app/src/main/kotlin/com/mikhailspeaks/masjidly/domain/NotificationSettings.kt
package com.mikhailspeaks.masjidly.domain

import kotlinx.serialization.Serializable

/** Mirrors iOS `NotificationSettings.swift`. */
@Serializable
data class NotificationSettings(
    var masterEnabled: Boolean = false,
    var adhanEnabled: Boolean = true,
    var iqamahEnabled: Boolean = true,
    var preAdhanReminderMinutes: Int? = null,
    var preIqamahReminderMinutes: Int? = null,
    var adhanFajr: Boolean = true,
    var adhanDhuhrJummah: Boolean = true,
    var adhanAsr: Boolean = true,
    var adhanMaghrib: Boolean = true,
    var adhanIsha: Boolean = true,
    var iqamahFajr: Boolean = true,
    var iqamahDhuhrJummah: Boolean = true,
    var iqamahAsr: Boolean = true,
    var iqamahMaghrib: Boolean = true,
    var iqamahIsha: Boolean = true,
)

enum class AdhanPrayerToggle(val label: String) {
    FAJR("Fajr"),
    DHUHR_JUMMAH("Dhuhr / Jummah"),
    ASR("Asr"),
    MAGHRIB("Maghrib"),
    ISHA("Isha"),
}

enum class IqamahPrayerToggle(val label: String) {
    FAJR("Fajr"),
    DHUHR_JUMMAH("Dhuhr / Jummah"),
    ASR("Asr"),
    MAGHRIB("Maghrib"),
    ISHA("Isha"),
}

fun NotificationSettings.adhanEnabled(prayer: AdhanPrayerToggle): Boolean = when (prayer) {
    AdhanPrayerToggle.FAJR -> adhanFajr
    AdhanPrayerToggle.DHUHR_JUMMAH -> adhanDhuhrJummah
    AdhanPrayerToggle.ASR -> adhanAsr
    AdhanPrayerToggle.MAGHRIB -> adhanMaghrib
    AdhanPrayerToggle.ISHA -> adhanIsha
}

fun NotificationSettings.setAdhanEnabled(prayer: AdhanPrayerToggle, enabled: Boolean) {
    when (prayer) {
        AdhanPrayerToggle.FAJR -> adhanFajr = enabled
        AdhanPrayerToggle.DHUHR_JUMMAH -> adhanDhuhrJummah = enabled
        AdhanPrayerToggle.ASR -> adhanAsr = enabled
        AdhanPrayerToggle.MAGHRIB -> adhanMaghrib = enabled
        AdhanPrayerToggle.ISHA -> adhanIsha = enabled
    }
}

fun NotificationSettings.iqamahEnabled(prayer: IqamahPrayerToggle): Boolean = when (prayer) {
    IqamahPrayerToggle.FAJR -> iqamahFajr
    IqamahPrayerToggle.DHUHR_JUMMAH -> iqamahDhuhrJummah
    IqamahPrayerToggle.ASR -> iqamahAsr
    IqamahPrayerToggle.MAGHRIB -> iqamahMaghrib
    IqamahPrayerToggle.ISHA -> iqamahIsha
}

fun NotificationSettings.setIqamahEnabled(prayer: IqamahPrayerToggle, enabled: Boolean) {
    when (prayer) {
        IqamahPrayerToggle.FAJR -> iqamahFajr = enabled
        IqamahPrayerToggle.DHUHR_JUMMAH -> iqamahDhuhrJummah = enabled
        IqamahPrayerToggle.ASR -> iqamahAsr = enabled
        IqamahPrayerToggle.MAGHRIB -> iqamahMaghrib = enabled
        IqamahPrayerToggle.ISHA -> iqamahIsha = enabled
    }
}

fun NotificationSettings.syncMasterFlag() {
    masterEnabled = adhanEnabled ||
        iqamahEnabled ||
        preAdhanReminderMinutes != null ||
        preIqamahReminderMinutes != null
}
